### PR TITLE
Fix: ensure custom session can be provided to rest client

### LINKED ISSUE #1395: Cannot pass session to rest client
The `_validate_session_raise_for_status` method is being called before the session is set. This crashes the program.

### PR DESCRIPTION
**Pull request by z3z1ma**
### Description

A simple fix for a small oversight in the rest client. Additionally a test proving you can pass a custom session and there is no crash.

### Related Issues

- Fixes #1395

### CODE REVIEW

**Comment by z3z1ma** [comment]
This is also confusing
https://github.com/dlt-hub/dlt/blob/64c953816c99643e1b6569f5b1fcba12dcc8de2b/dlt/sources/helpers/rest_client/client.py#L186-L190

Why is this there is raise for status is intended to be configured at session level

**Comment by z3z1ma** [comment]
Also note, you bypass `session.request` though this code path here
https://github.com/dlt-hub/dlt/blob/64c953816c99643e1b6569f5b1fcba12dcc8de2b/dlt/sources/helpers/rest_client/client.py#L92-L133


And call `session.send` directly BUT there is no retry logic on `send`. Its only monkey-patched into `session.request` here
https://github.com/dlt-hub/dlt/blob/64c953816c99643e1b6569f5b1fcba12dcc8de2b/dlt/sources/helpers/requests/retry.py#L242 

I can fix that too.

**Comment by burnash** [comment]
> you bypass session.request though this code path here

Yes, very good catch! Thanks for the fix again.

> This is also confusing
> ```
> def raise_for_status(response: Response, *args: Any, **kwargs: Any) -> None: 
>     response.raise_for_status() 
> if "response" not in hooks: 
>     hooks["response"] = [raise_for_status]
> ```

Agreed, not the most elegant solution, the idea is to enable hooks while having the rase_for_status as a default. This is also related to the session check. I wanted to give the user a way to override the default behaviour. Naturally the suggestions of how to do this better are very welcome!

**Review by rudolfix — APPROVED** [review]
LGTM!

one question: are you trying to use a session that is not from `requests` library? ie `httpx`?

**Review by rudolfix — CHANGES_REQUESTED** [review]
I think that your change somehow broke one test where I expect the host is mocked.
```
FAILED tests/sources/helpers/rest_client/test_client.py::TestRESTClient::test_oauth_jwt_auth_success - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.example.com', port=443): Max retries exceeded with url: /oauth/token (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f08d16f1710>: Failed to establish a new connection: [Errno -2] Name or service not known'))
```
@z3z1ma pls ping us if you have time to investigate that soon, otherwise we'll take over

**Comment by z3z1ma** [comment]
Test is passing for me 🤔  interesting, it shouldnt be flaky?

image



Also was doing my PR reply using https://github.com/pwntester/octo.nvim and totally replied via appending/editing your comments by accident (didnt even know that was possible)

**Comment by z3z1ma** [comment]
> LGTM!
> 
> one question: are you trying to use a session that is not from `requests` library? ie `httpx`?

Actually I was trying to use the dlt requests Client.session (I had done some custom stuff but wanted the default retry logic).


> Agreed, not the most elegant solution, the idea is to enable hooks while having the raise_for_status as a default. This is also related to the session check. I wanted to give the user a way to override the default behaviour. Naturally the suggestions of how to do this better are very welcome!

Would it make more sense to inherit from the session by default? This way the behavior is consistent within a source but can still be overridden on a per endpoint basis? I need to double check if/how sessions exposes this.

File: dlt/sources/helpers/rest_client/client.py [hunk 1]
```diff
@@ -90,15 +89,6 @@ def __init__(
 
         self.data_selector = data_selector
 
-    def _validate_session_raise_for_status(self, session: BaseSession) -> None:
-        # dlt.sources.helpers.requests.session.Session
-        # has raise_for_status=True by default
```
[the remark this answers is not in the document]
Comment (z3z1ma) [reply dlt/sources/helpers/rest_client/client.py:95]: Ok re-added

File: dlt/sources/helpers/rest_client/client.py [hunk 2]
```diff
@@ -80,8 +80,7 @@ def __init__(
         self.auth = auth
 
         if session:
-            self._validate_session_raise_for_status(session)
-            self.session = session
+            self.session = _warn_session_raise_for_status(session)
```
[the remark this answers is not in the document]
Comment (z3z1ma) [reply dlt/sources/helpers/rest_client/client.py]: Ok will update the function name to `_warn_and_return` along those lines

File: dlt/sources/helpers/rest_client/client.py [hunk 3]
```diff
@@ -296,3 +286,11 @@ def detect_paginator(self, response: Response, data: Any) -> BasePaginator:
                 " instance of the paginator as some settings may not be guessed correctly."
             )
         return paginator
+
+
+def _warn_session_raise_for_status(session: BaseSession) -> BaseSession:
```
[the remark this answers is not in the document]
Comment (burnash) [reply dlt/sources/helpers/rest_client/client.py]: Thanks for elaborating! Agreed.

**Comment by burnash** [comment]
> Would it make more sense to inherit from the session by default? This way the behavior is consistent within a source but can still be overridden on a per endpoint basis? I need to double check if/how sessions exposes this.

Yeah I was also contemplating this. But in that case how do we allow the user to pass their own Session?

**Comment by burnash** [comment]
> Test is passing for me 🤔 interesting, it shouldnt be flaky?

I've checked out the branch locally and also don't see this test_oauth_jwt_auth_success error. 🤔

**Comment by burnash** [comment]
> this is due to the OAuthAuth using default dlt session which is created before the mock by some other test... let me take a look and fix it

Interesting

**Review by burnash — APPROVED** [review]
Looks good, @z3z1ma! Thanks for fixing it!

**Pull request merged**